fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectory.java

```diff
@@ -0,0 +1,375 @@
+     * @return FDBLuceneFileReference
+     */
+    public CompletableFuture<FDBLuceneFileReference> getFDBLuceneFileReference(final String name) {
+        LOG.trace("getFDBLuceneFileReference {}", name);
+        FDBLuceneFileReference fileReference = this.fileReferenceCache.getIfPresent(name);
+        if (fileReference == null) {
+            return txn.get(metaSubspace.pack(name))
+                    .thenApplyAsync((value) -> {
+                            FDBLuceneFileReference fetchedref = value == null ? null : new FDBLuceneFileReference(Tuple.fromBytes(value));
+                            if (fetchedref != null) {
+                                this.fileReferenceCache.put(name, fetchedref);
+                            }
+                            return fetchedref;
+                        }
+                    );
+        } else {
+            return CompletableFuture.supplyAsync(() -> fileReference);
+        }
+    }
+
+    /**
+     * Puts a file reference in the meta subspace and in the cache under the given name.
+     * @param name name for the file reference
+     * @param reference the file reference being inserted
+     */
+    public void writeFDBLuceneFileReference(final String name, final FDBLuceneFileReference reference) {
+        LOG.trace("writeFDBLuceneFileReference {}", reference);
+        txn.set(metaSubspace.pack(name), reference.getTuple().pack());
+        fileReferenceCache.put(name, reference);
+    }
+
+    /**
+     * Writes data to the given block under the given id.
+     * @param id id for the data
+     * @param block block for the data to be stored in
+     * @param value the data to be stored
+     */
+    public void writeData(long id, int block, byte[] value) {
+        LOG.trace("writeData id={}, block={}, valueSize={}", id, block, value.length);
+        Verify.verify(value.length <= blockSize);
+        txn.set(dataSubspace.pack(Tuple.from(id, block)), value);
+    }
+
+    /**
+     * Reads known data from the directory.
+     * @param resourceDescription Description should be non-null, opaque string describing this resource; used for logging
+     * @param referenceFuture the reference where the data supposedly lives
+     * @param block the block where the data is stored
+     * @return Completable future of the data returned
+     * @throws IOException if blockCache fails to get the data from the block
+     * @throws NullPointerException if a reference with that id hasn't been written yet.
+     */
+    public CompletableFuture<byte[]> readBlock(String resourceDescription, CompletableFuture<FDBLuceneFileReference> referenceFuture, int block) throws IOException {
+        try {
+            LOG.trace("readBlock resourceDescription={}, block={}", resourceDescription, block);
+            final FDBLuceneFileReference reference = referenceFuture.join(); // Tried to fully pipeline this but the reality is that this is mostly cached after listAll, delete, etc.
```

Comment:
If like below we check if then reference is null then we can avoid the NPE that is thrown and instead consistently throw a NoSuchFile Exception if the resource doesn't exist.